### PR TITLE
BME-128 Optimising thread locking to avoid contention while fixing IO…

### PR DESCRIPTION
**Pull request by maddie-j**
Optimising the thread locking by trying to use read-write locks and threadsafe datastructures to improve throughput of threads. This doesn't entirely eliminate the hitching in game (as the cause of that is thread pool exhaustion) but it significantly reduces it, both in time and in frequency.

### CODE REVIEW

File: src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java [hunk 1]
```diff
@@ -39,7 +43,13 @@ public LayerRegionTile(StorageAccess.Internal storage, BlazeRegistry.Key<Layer>
     public void tryLoad() {
         if(file.exists()) {
             Profilers.FileOps.LAYER_READ_TIME_PROFILER.begin();
-            lock.lockPriority();
+
+            // Trying to greedily acquire the locks with `tryLock()` to minimise delay on game thread
+            // TODO: Figure out a "failure" case so the timed tryLock can be used instead to "give up"
+            // if it's waited too long, thereby not blocking the tick
+            if (!fileLock.readLock().tryLock()) fileLock.readLock().lock();
+            if (!imageLock.writeLock().tryLock()) imageLock.writeLock().lock();
```
Comment (LordFokas) [inline src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: what's going on here?

File: src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java [hunk 2]
```diff
@@ -230,7 +230,6 @@ private void setMouse(double mouseX, double mouseY) {
     @Override
     public void render(PoseStack stack, int i0, int i1, float f0) {
         float scale = (float) getMinecraft().getWindow().getGuiScale();
-        fillGradient(stack, 0, 0, this.width, this.height, 0xFF333333, 0xFF333333);
```
Comment (LordFokas) [inline src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java:233]: how does this fix things?

File: src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java (hunk 1 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: So, starting with docs ref for `tryLock`: https://docs.oracle.com/en%2Fjava%2Fjavase%2F17%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/locks/ReentrantLock.html#tryLock()

> a call to tryLock() will immediately acquire the lock if it is available, whether or not other threads are currently waiting for the lock. This "barging" behavior can be useful in certain circumstances, even though it breaks fairness.

So, what these two lines are both doing (with their own respective lock) is:
1) Try to acquire the lock with that "barging" behaviour. If the lock happens to be free but several threads want it at that instance, the `tryLock` will _always_ win out. That gives the thread some priority,
2) If the lock isn't free at that instance, `tryLock()` will return false instead of waiting to acquire the lock.
3) Negating that returned value, the if statement will go on to do a normal call to lock the thread that _will_ pause to wait for it.
4) By the end of the line, the thread will have acquired that lock exactly once.

This idiom is being used similar to the way suggested in the function below the one linked above.

Am not sure if the situation will come up much at all that the barging behaviour will actually be triggered, but I figured why not try anyway, given it's on the critical path?

File: src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java (hunk 2 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java:233]: Same as https://github.com/LordFokas/BlazeMap/pull/38

File: src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java (hunk 1 above)
Comment (LordFokas) [reply src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: That explains the try lock... but why lock read and write? Doesn't the RWLock also lock read if you acquire a lock for write? It should IIRC.

File: src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java (hunk 2 above)
Comment (LordFokas) [reply src/main/java/com/eerussianguy/blazemap/feature/maps/WorldMapGui.java:233]: ok I'll just leave it be for now

File: src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java (hunk 1 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: Yes, a _single_ RWLock will hold both its read and write locks. However, two _separate_ locks are being acquired here:
- The `fileLock` is getting read locked to protect the LRT file on disk from being changed while it's being read from.
- The `imageLock` is getting write protected to protect the representation of the LRT saved to the class object in the `image` variable from being read from or written to while it's being populated by the contents being read from file.

Each lock is protecting a different resource. This makes it far more efficient as we're only locking the shared resources we're actually needing to access, and only locking them down as much as we need to protect the integrity of that particular shared resource.

Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: This granularity of locking is what most of the commit focusing on locking is about. We always want to lock as little as we possibly can get away with, which means more granular locks and other strategies designed to protect each resource separately. We don't want to unnecessarily block another thread who's working with data that doesn't interact with the data our thread is working on!

Comment (LordFokas) [reply src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java:51]: oh, right, I totally missed that it was two different locks for some reason...

File: src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java [hunk 3]
```diff
@@ -13,49 +15,142 @@
...
+     * A chunk can only be marked dirty if already holding the write lock, however can be marked as
+     * clean from either the read or write lock (eg if cache has been read out to file to save).
+     *
+     * Note: As isDirty() is _not_ protected by the cache lock (though this.dirty is marked volatile to
+     * synchronise its value between threads), placement of calls to setDirty() need to be considered in the
+     * context of providing the most useful value to other threads checking this chunks current cleanliness status
+     * (knowing they won't then be able to act further on this chunk without first acquiring its lock)
+     */
+    protected void setDirty(boolean dirty) {
+        rLock.lock();
```
Comment (LordFokas) [inline src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:131]: Read Lock is unnecessary because we already enforce Write Lock is being held, right?
So, if we're already holding the write lock, it's impossible for anyone to lock for Read until we let go of Write.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java [hunk 4]
```diff
@@ -65,44 +160,60 @@ public Persisted(RegionMDCache parent) {
...
-                BlazeRegistry.Key<DataType> key = (BlazeRegistry.Key<DataType>) (Object) BlazeMapAPI.MASTER_DATA.findOrCreate(stream.readResourceLocation());
+            // Yes I see the irony that the read() function needs a write lock and the write() function needs a read lock
+            wLock.lock();
+            try {
+                for(int i = 0; i < entries; i++) { // Why can't I just write Key<DataType<>> and have the compiler trust me? ffs.
+                    BlazeRegistry.Key<DataType> key = (BlazeRegistry.Key<DataType>) (Object) BlazeMapAPI.MASTER_DATA.findOrCreate(stream.readResourceLocation());
+
+                    if (key.value() == null) {
+                        BlazeMap.LOGGER.warn("Unrecognised DataType in MD cache. Did you uninstall a Blaze Map extension? Skipping cache read");
+                        throw new IOException("Cannot deserialize unregistered MD DataType");
```
Comment (LordFokas) [inline src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:177]: We should have a note somewhere that this is a poor solution and needs to be revisited.
But I don't actually require changes to this code in this PR. We just need to get to it later.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java [hunk 5]
```diff
@@ -1,67 +1,76 @@
...
-        return lock;
-    }
-
     public boolean isDirty() {
-        return dirty;
+        for (ChunkMDCache.Persisted chunk : chunks) {
+            if (chunk != null && chunk.isDirty()) {
+                return true;
+            }
+        }
```
Comment (LordFokas) [inline src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java:40]: I have issues.
On one hand, this is correct.
On the other hand, Chunk caches were passing their dirty state to their parent Region cache which was being kept locally so that we don't need to check 1024 chunk caches every time... that's why it was there in the first place.

Now, I could be wrong about the impact of these approaches, but my educated guess was that it'd be more performant to just keep a local value at the region level.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java (hunk 3 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:131]: So this is why I put in that big comment at the top there, because this is _not_ that normal scenario. 

Note specifically (emphasis added):
>  A chunk can only be marked dirty if already holding the write lock, however **can be marked as clean from either the read or write lock** (eg if cache has been read out to file to save).

In other words, this function is being allowed to be called from a read lock _in addition to_ a write lock, as long as the read lock is _only_ setting the `dirty` state to `false` (ie `clean`). This is so that, once the cache has been read out to the stream that's writing to disk (under a read lock for the cache), the `dirty` status can be updated to reflect this _without_ having to unnecessarily swap to a write lock.

The check below makes sure the only value it's ever allowed to be set to from a non-write lock is `clean`, given no actual changes to the cache contents are needed for that to be the case, and will throw an exception calling you out if you tried to dirty the cache without holding the write lock first.

Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:131]: I'm only grabbing the lock again to make sure that there is definitely _any_ lock being held at all and the dev didn't forget to lock completely. We don't want threads to be setting the variable to different values at the same time! 

And even though it's allowed to be written to under only a read lock, there's only _one_ value that it can be changed to under a read lock, so the worst that can happen is multiple threads simultaneously setting `dirty` to `false`. But if trying to change to `true`, a write lock must already be held from the calling function, and a write lock is an exclusive lock so there can be no contention from different threads trying to change the variable to a different value at the same time.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java (hunk 4 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:177]: Have created a ticket: https://blazemap.atlassian.net/browse/BME-188

File: src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java (hunk 5 above)
Comment (maddie-j) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java:40]: This was a trade off I did spend time looking into, because I really _didn't_ want to have to read 1024 chunks. But the trade-off I found was worth it.

Readin back on my notes from the time (https://discord.com/channels/709852148170620948/1267059848705605673/1267530673053241384), TL;DR:
>  I think minimising frequent thread synchronisation is more important than the occasional check for cleanliness before writing to a file max every 30s

The longer version:
The problem here is that we're dealing with three different optimisation goals here, and it's only possible to satisfy 2 of the 3:
- Best chunk access performance
- Best dirty check performance
- Thread safety

Previously, the code was optimising for "dirty check performance" and "thread safety". However, this was at the cost of at the cost of "chunk access performance". This is because, if _any_ chunk was being accessed, **every other chunk had to wait for it to be done**. Normally this is fine as the operation is generally fast, but it was _the_ major cause of the hitching if one thread tried to load a chunk from a region that needed to be read from disk. But the dirty check was fast, and it was always accurate as no chunks could be changed while writing them to disk.

In the new version of the code, I changed the locks so each chunk was locked separately. This meant that the game only hitched if _every_ thread was waiting for a chunk that was waiting on disk IO, as otherwise the threads that were looking up chunks in already loaded regions were not impacted at all by the other threads waiting on IO.

However, when it came to the `dirty` check, I then had to optimise between "thread safety" or "dirty check performance". _If_ I were to similarly use a variable on the region to optimise for "dirty check performance", this _could not_ be thread safe:
- When writing to disk, each chunk is only locked while it personally is being written. That means the chunk can be changed _after_ it's written to disk. If the `dirty` check is held on the region and applied after the file is written to disk, the updates to that chunk will _not_ be recognised. 
- Alternatively, if all chunks are locked while writing to disk to make sure that `dirty` check is accurate at the end of writing, that again causes unnecesary synchronisation on both reads and writes.
- If instead the variable is updated _before_ writing, to allow it to be updated by any chunks changed after they've been written out, this still won't be accurate if a chunk is changed before the write process gets to it but is then still output in its final state to disk. 
- Additionally, if the variable is updated before writing, it will be left as `clean` if an exception causes the write to fail, again not reflecting the true status of the region.
- I did even think of using the bits in a `long` to represent each chunk, with `1` meaning dirty and `0` meaning clean. That way, `dirty >  0` means the chunk is dirty. But it would have needed multiple `long`s to work, and those longs would have had to be synchronised, and it wasn't a clean solution.

Alternatively, I looked at where that dirty check was taking place. The dirty check is _only_ called before writing the region MD to file. This should happen at most every 30s, and should be only on the data threads. Given the rarity of this check and the fact it shouldn't impact the render thread, I considered it the least important of the three optimisation variables in the trade off. 

But also, that's why the complication around the locking state of the `dirty` variable on each chunk. Yes the region has to read 1024 variables, but it _never_ has to wait for a lock. The variable is `volatile`, so its value should be up to date in the CPU cache before reading. This is as fast as I can make accessing the `dirty` vairable in all chunks from a region while still protecting its access behind a function call (so it can't be written to without a lock).

In other words, I had to make a trade off, and I chose the set of trade offs that maximised average performance and correctness. And that's why it takes longer to check if a region is dirty in this new code.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java (hunk 3 above)
Comment (LordFokas) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java:131]: Ah, very well.

File: src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java (hunk 5 above)
Comment (LordFokas) [reply src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java:40]: I am satisfied
